### ISSUE TITLE
Improving application limited tests for BBR

### ISSUE THREAD

**Issue by alexrabi** [private-octopus/picoquic#1812]
As discussed in #1758, the current "application limited" test with BBR is insufficient as it sometimes mistakes busty traffic for path capacity saturation. This can lead to a very poor throughput. Testing the congestion window (`bytes_in_flight < cwnd`) is not enough as BBR uses the pacing rate as the main control variable, with the congestion window as a "backup". In normal operation, pacing dominates and the bytes in flight remain below the congestion window. Testing on the pacing rate is also imprecise, because pacing is implemented with a leaky bucket. There will be brief period in which the leaky bucket will be full, letting packets go without pacing, only limiting the last packet out of a train -- but that can happen whether the application is "limiting" or not.

We need to come up with a solution to more accurately detect whether we are application limited or not.

**Comment by alexrabi** [comment]
I'm wondering if it is worth taking a [few notes](https://queue.acm.org/detail.cfm?id=2209336) from the bufferbloat community in regards to this. The issue is being able to distinguish "good queues", i.e. queues that "absorb" bursts, but that eventually drain, from "bad queues", i.e. persistent, standing queues. If we are application limited, then the leaky bucket should act more like a good queue, rather than a bad queue.

The CoDel AQM algorithm uses the sojourn times of each packet to track the (local) minimum queue delay that packets experience. As long as the minimum queue delay is below some target threshold for some time interval, the queue is considered a good queue. Perhaps we can do something similar here to help identify application limited scenarios?

**Comment by huitema** [comment]
Damn, 4 months without a comment...

The key insight in the CoDel design is to distinguish between standing queues and transient queues, and measuring the "minimum" value of the queue rather than its maximum. I am wondering how to apply that concept in picoquic, as a typical high speed implementation will use the "just in time" API, and thus does not rely on the picoquic code to maintain queues. On the other hand, the just in time API allows implementations to mark streams or the datagram queue as "active", as in "there is data to send". The first approximation of "empty queue" would thus be "there is no data to send", and the code in `picoquic_prepare_stream_and_datagrams` (lines 2751 to 2941 in `sender.c`) manages that. There is no data to send if that function does not queue data, and also does not set the "more data" bit to indicate that there is data but it will not fit in the current packet.

The current code in sender.c will set the `last_sender_limited_time1 if there was no data to send. (The "preemptive repeat" flag instructs tested here instructs that stack to repeat old data if there is nothing else to send, in which case the sender is never limited.) This is part of the following block of code in `sender.c`:
~~~
                            if (length <= header_length) {
                                /* Mark the bandwidth estimation as application limited */
                                path_x->delivered_limited_index = path_x->delivered;
                                /* Notify the peer if something is blocked */
                                bytes_next = picoquic_format_blocked_frames(cnx, &bytes[length], bytes_max, &more_data, &is_pure_ack);
                                length = bytes_next - bytes;
                            }

                            if (no_data_to_send) {
                                path_x->last_sender_limited_time = current_time;
                            }
~~~
The BBR code only depends on the `last_sender_limited_time` in the startup phase, and only if the startup uses Hystart instead of the native BBR mechanism. In the other cases, BBR depends on `path_x->delivered_limited_index`. The code tests that when processing an acknowledgement, and will signal that in the `ack_state` variable passed to BBR (ack_state.is_app_limited). All that is fairly convoluted, which would be reason enough to try simplify it. The good news is that the test only fires if both pacing and CWIN would have allowed a packet to be sent, which means it will work for both Cubic using CWIN, and BBR using a combination of CWIN and pacing.

The main issue that I see is that the code fires the "app_limited" condition too often. Suppose the sender is "time limited" at time T, but sends data at time T+1 microsecond. The application is in fact keeping the bottleneck full, and is not `app_limited`. But on the other hand, setting app_limited two often only prevents BBR from reducing the bottleneck rate, and is not our problem.

@alexrabi is there a specific change that you would like to try?

**Comment by alexrabi** [comment]
Indeed, the "just in time" API complicates the problem, since the queue would be on the user application side of things, rather than in picoquic itself. As you say, if the stream and/or datagram queues are marked as active it makes sense for that to be interpreted as a persistent queue and therefore non-limited scenario. However, I feel like it is a bit too naïve to interpret a failure to queue data necessarily as a signal for "application limitation". 

I'll need to mull it over a bit, since it is a complex scenario, but here are some initial thoughts: 
The signalling between the application and the picoquic stack needs to be improved, since a lot of information that could be useful to picoquic will be lost when running the "just in time" API (e.g. the status of queues on the application side). For example, when marking a stream or datagrams ready to send, the application might want to signal that it is _almost_ ready to send. 

Signalling could also be further improved by allowing the application to signal roughly _how much_ data it is/is going to be ready to send on a stream or over datagrams. That would allow picoquic to glean _some_ information about how much queueing is going on on the application side, and could therefore act accordingly. E.g. something like:
`int picoquic_mark_datagram_ready_ex(picoquic_cnx_t* cnx, int is_ready, size_t bytes_to_send)` and 
`int picoquic_mark_active_stream_ex(picoquic_cnx_t* cnx, uint64_t stream_id, int is_active, void* v_stream_ctx, size_t bytes_to_send)`

**Comment by huitema** [comment]
@alexrabi please look at PR #1903, which introduces a process for letting application control the per stream flow control. Would that fix your issue?

**Comment by victorstewart** [comment]
i tried to generate a synthetic packet flow to trigger this behavior but haven't been successful yet. if it happens again and you share a packet capture that would go a long way to figuring out how to generate a reproducer, to then test any code mitigations against. 

or if you have any ideas @huitema i can try them out.

**Comment by alexrabi** [comment]
> [@alexrabi](https://github.com/alexrabi) please look at PR [#1903](https://github.com/private-octopus/picoquic/pull/1903), which introduces a process for letting application control the per stream flow control. Would that fix your issue?

Sorry for the several months late response (yikes)! This is a step in the right direction for sure. This solution appears to be limited to streams however, so it does not fully fix this issue (i.e. it does not account for applications that primarily use datagrams, and not streams).

**Comment by huitema** [comment]
@alexrabi datagrams in QUIC are not subject to flow control. If the application uses datagrams, it needs to solve flow control problems insides the application protocol.

**Comment by alexrabi** [comment]
@huitema Indeed, there is no flow control for datagrams, but my observations indicate that the BBR CCA  does not handle such scenarios well (I believe the pacing rate specifically may be affected when the application does not send enough data).

**Comment by huitema** [comment]
Yes, that's a known issue with BBR. BBR will measure data rate for each successive RTT. If the measure is lower than the current rate, it will lower that rate, unless it recognizes that the application is "limited". But the test of whether the application is limited is a bit fuzzy, so BBR will sometime wrongly decide that it is not limited and then reduce the bandwidth. Or, there could be a temporary network event that causes a slowdown. Batch applications will recover because they send more data, but a limited application can only measure a data rate equal at most to its send rate.

**Comment by alexrabi** [comment]
Yeah, that description is in line with what I have observed. If the application is not _consistently_ greedy for resources, for example, having a period of  high utilization, followed by a period of low utilization, and then another period of high utilization, then the pacing rate may be reduced in the low utilization period, and does not recover for the second period of high utilization (since the sending is now limited by the low pacing rate), leading to poor overall performance. One issue is that there is a disconnect between what is known at the application and what is known by the CCA. Maybe there could be a way for the application to signal that it is entering a state where low utilization is expected, allowing the application limited test to act accordingly?

It's worth noting that the scenario that I have been looking at is quite complex. The application in my case is using picoquic as a tunnel, which greatly exacerbates the issue. In other words, there is the possibility of CC/CC, which can have a myriad of interactions. In this case, I believe that if picoquic's BBR erroneously interprets the situation as not being application limited, thus lowering the pacing rate, then the inner CCA may accept the reduced bandwidth from the pacing rate as the actual bandwidth on the line. That could create a sort of feedback loop where the application is not sending enough data because picoquic's pacing rate is too low, and picoquic's pacing rate will not recover because the application is not sending enough data.

**Comment by huitema** [comment]
That issue is one of my motivation for developing C4 (https://github.com/private-octopus/picoquic/c4) -- a design principle is to not reduce the bandwidth unless there is actual congestion.

**Comment by alexrabi** [comment]
Sounds like a very interesting initiative! So the idea of C4 is to eventually develop a new CCA for real-time applications which would take issues like this into account then?

**Comment by huitema** [comment]
Yes, that's the idea. Other motivation include working on "every" path, including low quality Wi-Fi that can cause huge delay jitter.

**Comment by alexrabi** [comment]
Very interesting indeed. Perhaps some of the lessons learned could be ported to other CCA's as well!

**Comment by joynwang** [comment]
Hi,  I am new here. I have been testing BBRv1, BBRv3 and Cubic with this implementation. It seems that I encounter app-limited issue. I have dumbell scenario with bottleneck link to be 5Gbps (the rest with 100Gbps). The client sends 1GB data to the server, but the throughput is stuck at 300Mbps (Far from 5Gbps).  I wonder if you know the cause or the solution to this problem. Thanks.

**Comment by huitema** [comment]
@joynwang could you please open a separate issue? Also, please specify in the title the environment that you are using. Is this the NS3 simulator?

**Comment by joynwang** [comment]
﻿A new issue created: https://github.com/private-octopus/picoquic/issues/2089 <1ddc9760-474a-4e1b-9369-a99cbcb3190a>


I appreciate it any insight on this. Thanks. 


Joy
Get Outlook for Mac <margin-top: 0px; margin-bottom: 0px;> 

From: Christian Huitema ***@***.***>
Date: Friday, April 10, 2026 at 4:17 PM
To: private-octopus/picoquic ***@***.***>
Cc: Wang, Joy - 0665 - MITLL ***@***.***>, Mention ***@***.***>
Subject: [EXT] Re: [private-octopus/picoquic] Improving application limited tests for BBR (Issue #1812)


This Message Is From an External Sender
This message came from outside the Laboratory.



huitema left a comment (private-octopus/picoquic#1812) <3f621914-f049-4148-9e27-b89d7dce02c7>
@joynwang <margin-top: 0px; margin-bottom: 0px;> could you please open a separate issue? Also, please specify in the title the environment that you are using. Is this the NS3 simulator?
